### PR TITLE
types(firestore): make `withConverter` retro compatible

### PR DESCRIPTION
**Pull request by posva**
I found this while migrating vuefire to Firebase 10: it was no longer possible to write:

```ts
itemRef.withConverter<number>({
  // ...
})
```

As it would fail in TypeScript. I think adding the default type value here helps people migrate to Firebase 10.
I haven't added (or even checked where to add) a test because I'm unaware of any discussion that happened for this.

It's also possible that other places would benefit from a similar change of adding a default TS parameter

### CODE REVIEW

**Comment by ehsannas** [comment]
Thanks for your contribution @posva . 

@dconeybe PTAL

**Comment by dconeybe** [comment]
Thank you for the contribution! Could you show a complete code sample that used to compile in v9 and doesn't compile in v10? I'd like to copy/paste it and see what other places could benefit.

**Comment by posva** [comment]
Sure:

```ts
import { collection, getFirestore, type DocumentData } from 'firebase/firestore'

const db = getFirestore()
collection(db, 'todos').withConverter<
  { text: string; finished: boolean },
  DocumentData
>({
  fromFirestore: (snapshot) => {
    const data = snapshot.data()
    // ... do actual checks
    return { text: data.text, finished: data.finished }
  },
  toFirestore: (todo) => todo,
})
```

The 2nd generic is now required but it wasn't before. `itemRef` could also be a `doc()` (from the original post)

**Comment by posva** [comment]
Hello! any news @dconeybe ? The later this is merged the less value is has as the point is to ease the upgrade from firebase 9 to 10 for users

File: packages/firestore/src/lite-api/reference.ts
Comment (dconeybe) [inline packages/firestore/src/lite-api/reference.ts:1]: Please add a changeset so that a release notes entry will be included for this feature.

To do this:
1. run `yarn changeset` in the lop-level directory of the git repository.
2. When asked "Which packages would you like to include?" select "@firebase/firestore" and "firebase"
3. When asked "Which packages should have a major bump?" do not select any.
4. When asked "Which packages should have a minor bump?" select both "all packages"
5. For "Summary" enter something like this: "Added a default template type parameter to withConverter() functions to improve backwards compatibility with the v9 SDK."
6. When asked "Is this your desired changeset?" press Y
7. Commit the file created in the `.changeset` directory to git and push it up.

**Review by dconeybe — DISMISSED** [review]
Thank you for your contribution, and apologies for the delay in providing feedback. After discussing with the team, we have decided to incorporate your suggestion. Since this change involves a public API change there is some additional process that we need to go through to get it approved; however, I do not expect any issues. But it may be 1-3 weeks to get approval to merge it. Also note that we have incorporated this suggestion into our node server sdk (https://github.com/googleapis/nodejs-firestore/pull/1887), which has less a less stringent process for public api changes. In the interim, I've left a bit of feedback. Please take a look. Thanks again for contributing to Firestore!

File: packages/firestore/src/lite-api/reference.ts
Comment (posva) [reply packages/firestore/src/lite-api/reference.ts:1]: BTW feel free to take it from here, any maintainer should be able to push any changes they please.

I will try to come back to this in a couple of weeks

Comment (dconeybe) [reply packages/firestore/src/lite-api/reference.ts:1]: Sounds good. I've implemented my suggested changes. We can take it from here. Just a reminder that it will probably be a couple of weeks before this gets merged due to the extra process involved with public API changes.

**Comment by dconeybe** [comment]
I'm looking into the GitHub Actions failures of "Deploy Project Config" and "Update API reports". It appears that they may need to be adjusted to support external contributions.

**Review by tom-andersen — APPROVED** [review]

**Comment by tom-andersen** [comment]
@markarndt I think we need your approval on changeset before we can merge. Could you take a look?

**Review by markarndt — APPROVED** [review]

**Comment by tom-andersen** [comment]
@hsubox76 I think I need your approval.

**Review by hsubox76 — APPROVED** [review]

**Pull request merged**
